internal/component/mimir/rules/kubernetes/events.go

```diff
@@ -4,108 +4,156 @@ import (
 	promv1 "github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1"
+	promListers "github.com/prometheus-operator/prometheus-operator/pkg/client/listers/monitoring/v1"
 	"github.com/prometheus/prometheus/model/rulefmt"
+	"k8s.io/apimachinery/pkg/labels"
+	coreListers "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/util/workqueue"
 	"sigs.k8s.io/yaml" // Used for CRD compatibility instead of gopkg.in/yaml.v2
 )
 
-const eventTypeSyncMimir kubernetes.EventType = "sync-mimir"
+const (
+	eventTypeSyncMimir kubernetes.EventType = "sync-mimir"
+)
+
+type healthReporter interface {
+	reportUnhealthy(err error)
+	reportHealthy()
+}
+
+type eventProcessor struct {
+	queue    workqueue.RateLimitingInterface
+	stopChan chan struct{}
+	health   healthReporter
+
+	mimirClient       mimirClient.Interface
+	namespaceLister   coreListers.NamespaceLister
+	ruleLister        promListers.PrometheusRuleLister
+	namespaceSelector labels.Selector
+	ruleSelector      labels.Selector
+	namespacePrefix   string
 
-func (c *Component) eventLoop(ctx context.Context) {
+	metrics *metrics
+	logger  log.Logger
+
+	currentState    kubernetes.RuleGroupsByNamespace
+	currentStateMtx sync.RWMutex
+}
+
+func (e *eventProcessor) run(ctx context.Context) {
 	for {
-		eventInterface, shutdown := c.queue.Get()
+		eventInterface, shutdown := e.queue.Get()
 		if shutdown {
-			level.Info(c.log).Log("msg", "shutting down event loop")
+			level.Info(e.logger).Log("msg", "shutting down event loop")
 			return
 		}
 
 		evt := eventInterface.(kubernetes.Event)
-		c.metrics.eventsTotal.WithLabelValues(string(evt.Typ)).Inc()
-		err := c.processEvent(ctx, evt)
+		e.metrics.eventsTotal.WithLabelValues(string(evt.Typ)).Inc()
+		err := e.processEvent(ctx, evt)
 
 		if err != nil {
-			retries := c.queue.NumRequeues(evt)
+			retries := e.queue.NumRequeues(evt)
 			if retries < 5 {
-				c.metrics.eventsRetried.WithLabelValues(string(evt.Typ)).Inc()
-				c.queue.AddRateLimited(evt)
-				level.Error(c.log).Log(
+				e.metrics.eventsRetried.WithLabelValues(string(evt.Typ)).Inc()
+				e.queue.AddRateLimited(evt)
+				level.Error(e.logger).Log(
 					"msg", "failed to process event, will retry",
 					"retries", fmt.Sprintf("%d/5", retries),
 					"err", err,
 				)
 				continue
 			} else {
-				c.metrics.eventsFailed.WithLabelValues(string(evt.Typ)).Inc()
-				level.Error(c.log).Log(
+				e.metrics.eventsFailed.WithLabelValues(string(evt.Typ)).Inc()
+				level.Error(e.logger).Log(
 					"msg", "failed to process event, max retries exceeded",
 					"retries", fmt.Sprintf("%d/5", retries),
 					"err", err,
 				)
-				c.reportUnhealthy(err)
+				e.health.reportUnhealthy(err)
 			}
 		} else {
-			c.reportHealthy()
+			e.health.reportHealthy()
 		}
 
-		c.queue.Forget(evt)
+		e.queue.Forget(evt)
 	}
 }
 
-func (c *Component) processEvent(ctx context.Context, e kubernetes.Event) error {
-	defer c.queue.Done(e)
+func (e *eventProcessor) stop() {
+	close(e.stopChan)
```

Comment:
It feels a bit weird to have `stopChan` inside `eventProcessor`, given that it doesn't actually do anything for `eventProcessor`. Maybe there is a better way to alert other code when things are being stopped.

It's also notable that `stopChan` is close without waiting for `run` to complete, and maybe this should be documented.

Reply:
The stop channel is how the Kubernetes components are told to shutdown. I moved the stop channel to `eventProcessor` because it "owns" the Kubernetes components after they are created and so should be responsible for telling them to shutdown.

Reply:
> It's also notable that `stopChan` is close without waiting for `run` to complete, and maybe this should be documented.

Right but `e.queue.ShutdownAndDrain()` waits for all events to be drained from the queue and then causes subsequent calls to `e.queue.Get()` to return `true` for the shutdown value which will stop `run`.

Reply:
Added some comments that clarify this behavior in f46d88eee49adc9d7a7d670bad16bd2ca3bdb6af